zend_abstract_interface/config/config_decode.c

```diff
@@ -0,0 +1,201 @@
+#include "config_decode.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <main/php.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+
+#include "config.h"
+
+static bool zai_config_decode_bool(zai_string_view value, zval *decoded_value) {
+    if (value.len == 1) {
+        if (strcmp(value.ptr, "1") == 0) {
+            ZVAL_TRUE(decoded_value);
+        } else if (strcmp(value.ptr, "0") == 0) {
+            ZVAL_FALSE(decoded_value);
+        }
+    } else if ((value.len == 4 && strcasecmp(value.ptr, "true") == 0)) {
+        ZVAL_TRUE(decoded_value);
+    } else if ((value.len == 5 && strcasecmp(value.ptr, "false") == 0)) {
+        ZVAL_FALSE(decoded_value);
+    }
+    return Z_TYPE_P(decoded_value) == IS_TRUE || Z_TYPE_P(decoded_value) == IS_FALSE;
+}
+
+/* Since strtod() supports conversion of constants like 'NaN' and 'INF', as well
+ * as successfully decoding values such as '4foo', we check that only numbers,
+ * decimal points, and whitespace exist before passing to strtod().
+ */
+static bool zai_config_is_valid_double_format(const char *str) {
+    bool seen_decimal = false;
+    while (*str) {
+        if (!isdigit(*str) && !isspace(*str) && *str != '.') return false;
+        if (*str == '.') {
+            /* strtod will convert '4.2.0' to a double, so we block it. */
+            if (seen_decimal) return false;
+            seen_decimal = true;
+        }
+        str++;
+    }
+    return true;
+}
+
+static bool zai_config_decode_double(zai_string_view value, zval *decoded_value) {
+    if (!zai_config_is_valid_double_format(value.ptr)) return false;
+
+    char *endptr = (char *)value.ptr;
+
+    // The strtod function is a bit tricky, so I've quoted docs to explain code
+
+    /* Since 0 can legitimately be returned on both success and failure, the
+     * calling program should set errno to 0 before the call, and then
+     * determine if an error occurred by checking whether errno has a nonzero
+     * value after the call.
+     */
+    errno = 0;
+    double result = strtod(value.ptr, &endptr);
```

Comment:
This is locale sensitive. Do not use strtod in PHP code unless you actually want things differently handled by locale. (e.g. the comma vs dot as decimals separator.)
`zend_strtod` handles this properly for you.

Reply:
Very good point. Although `strtod` is what we're currently using to parse doubles, I don't see any reason why we would not want to change this to the Zend flavor.